go/gas-oracle/oracle/updater_interface.go

```diff
@@ -0,0 +1,158 @@
+package oracle
+
+import (
+	"context"
+	"errors"
+	"math"
+	"math/big"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/go/gas-oracle/bindings"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// getLatestBlockNumberFn is used by the GasPriceUpdater
+// to get the latest block number. The outer function binds the
+// inner function to a `bind.ContractBackend` which is implemented
+// by the `ethclient.Client`
+func wrapGetLatestBlockNumberFn(backend bind.ContractBackend) func() (uint64, error) {
+	return func() (uint64, error) {
+		tip, err := backend.HeaderByNumber(context.Background(), nil)
+		if err != nil {
+			return 0, err
+		}
+		return tip.Number.Uint64(), nil
+	}
+}
+
+// DeployContractBackend represents the union of the
+// DeployBackend and the ContractBackend
+type DeployContractBackend interface {
+	bind.DeployBackend
+	bind.ContractBackend
+}
+
+// updateL2GasPriceFn is used by the GasPriceUpdater
+// to update the L2 gas price
+// perhaps this should take an options struct along with the backend?
+// how can this continue to be decomposed?
+func wrapUpdateL2GasPriceFn(backend DeployContractBackend, cfg *Config) (func(float64) error, error) {
+	if cfg.privateKey == nil {
+		return nil, errNoPrivateKey
+	}
+	if cfg.chainID == nil {
+		return nil, errNoChainID
+	}
+
+	opts, err := bind.NewKeyedTransactorWithChainID(cfg.privateKey, cfg.chainID)
+	if err != nil {
+		return nil, err
+	}
+	// Once https://github.com/ethereum/go-ethereum/pull/23062 is released
+	// then we can remove setting the context here
+	if opts.Context == nil {
+		opts.Context = context.Background()
+	}
+
+	// Create a new contract bindings in scope of the updateL2GasPriceFn
+	// that is returned from this function
+	contract, err := bindings.NewGasPriceOracle(cfg.gasPriceOracleAddress, backend)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(num float64) error {
+		log.Trace("UpdateL2GasPriceFn", "gas-price", num)
+		if cfg.gasPrice == nil {
+			// Set the gas price manually to use legacy transactions
+			gasPrice, err := backend.SuggestGasPrice(context.Background())
+			if err != nil {
+				log.Error("cannot fetch gas price", "message", err)
+				return err
+			}
+			log.Trace("fetched L2 tx.gasPrice", "gas-price", gasPrice)
+			opts.GasPrice = gasPrice
+		} else {
+			// Allow a configurable gas price to be set
```

Comment:
TODO: add a polling cache to the sync service that allows through 0 gas price txs iff it is the owner of the `OVM_GasPriceOracle`